File: DemoServer/PythonDemo/demo_methods/Advanced/Subscriptions.py
from demo.holder import DocumentStoreHolder
from pyravendb.subscriptions.data import SubscriptionCreationOptions, SubscriptionWorkerOptions
from pyravendb.subscriptions.data import SubscriptionOpeningStrategy
from pyravendb.custom_exceptions.exceptions import SubscriptionClosedException


class Subscriptions(object):

    @staticmethod
    def subscriptions(subscriptionName=None):
        companies = []
        with DocumentStoreHolder.get_store() as store:
            if not subscriptionName:
                subscription_creation_options = SubscriptionCreationOptions("From Companies")
                subscriptionName = store.subscriptions.create(subscription_creation_options)

            worker_options = SubscriptionWorkerOptions(subscriptionName, strategy=SubscriptionOpeningStrategy.take_over,
                                                       close_when_no_docs_left=True)
            with store.subscriptions.get_subscription_worker(worker_options) as subscription_worker:
                try:
                    subscription_worker.run(
                        partial(lambda batch, x=companies: x.extend([item.raw_result for item in batch.items]))).join()
                except SubscriptionClosedException:
                    # That's expected
                    pass

        return companies

File: DemoServer/Demos/Advanced/Subscriptions.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using DemoServer.Controllers;
using DemoServer.Entities;
using DemoServer.Helpers;
using Microsoft.AspNetCore.Mvc;
using Raven.Client.Documents.Subscriptions;
using Raven.Client.Exceptions.Documents.Subscriptions;

namespace DemoServer.Demos.Advanced
{
    public partial class AdvancedController : BaseController
    {
        [HttpGet]
        [Route("/advanced/subscriptions")]
        [Demo("Subscriptions", demoOrder: 230)]
        public async Task<object> Subscriptions(string subscriptionName = null)
        {
            var companies = new List<object>();
            if (subscriptionName == null)
            {
                var subscriptionCreationOptions = new SubscriptionCreationOptions
                {
                    Query = @"From Companies"
                };
                subscriptionName = DocumentStoreHolder.Store.Subscriptions.Create(subscriptionCreationOptions);
            }
            var workerOptions = new SubscriptionWorkerOptions(subscriptionName)
            {
                Strategy = SubscriptionOpeningStrategy.TakeOver,
                CloseWhenNoDocsLeft = true
            };
            using (var subscriptionWorker = DocumentStoreHolder.Store.Subscriptions.GetSubscriptionWorker(workerOptions))
            {
                try
                {
                    await subscriptionWorker.Run(batch =>
                    {
                        foreach (var item in batch.Items)
                        {
                            companies.Add(item.Result);
                        }
                    });
                }
                catch (SubscriptionClosedException)
                {
                    // That's expected
                }

                return companies;
            }
        }
    }
}
